app/main/views.py

```diff
@@ -104,21 +104,28 @@ def update(service_id, section):
         list(request.form.items()) + list(request.files.items())
     )
 
-    errors = Validate(content, service, posted_data, s3_uploader).errors
+    form = Validate(content, service, posted_data, s3_uploader)
+    update = {}
 
     for question_id in posted_data:
-        if question_id not in errors:
-            service_loader.set(service, question_id, "new value")
-
-    service_loader.post(service)
+        if question_id not in form.errors and question_id in form.clean_data:
+            update[question_id] = form.clean_data[question_id]
+
+    if update:
+        service_loader.post(
```

Comment:
In both cases if the API update fails we end up with unused files in S3. It's actually better right now since if you replace a pdf with a pdf the URL stays the same so the API update doesn't actually change anything except creating an archived copy.

At any rate, it seems that it would be hard to consistently recover from this (e.g. if API call failed due to network and so will the S3 delete requests), but we can have a batch task looking for files not referenced in services if this ever becomes a storage/cost problem.